binary_search_tree/BinarySearchTree.py

```diff
@@ -0,0 +1,334 @@
+class Node():
+    # This class holds an actual Node in our BST
+
+    # the actual data which is stored in the node
+    __data = None
+    # left child of the node
+    __left = None
+    # right child of the node
+    __right = None
+
+    def __init__(self, data=None, left=None, right=None):
+        # here self.data, self.left, self.right are not variables
+        # these are setters which are defined below
+        self.data = data
+        self.left = left
+        self.right = right
+
+    @property
+    def data(self):
+        # adding @property makes this a getter and can be accessed by ObjectName.data
+
+        return self.__data
+
+    @data.setter
+    def data(self, value):
+        # we need setter for "__data" since its a private variable
+        # Use of setters is to validate data or perform some action before assiging the data
+        # we only want our data to be a number/int
+        if isinstance(value, int):
+            self.__data = value
+        else:
+            self.__data = None
+            print("Invalid data: {} only integers(int) are accepted.".format(value))
+
+    # now we will repeat the same for "left" and "right" variables too
+    @property
+    def left(self):
+        return self.__left
+
+    @left.setter
+    def left(self, value):
+        # since left is not just any variable but a node itself
+        if (isinstance(value, Node)) or (value is None):
+            self.__left = value
+        else:
+            print("Invalid left node, should be an instance of class 'Node'.")
+
+    # repeat the same for right node
+    @property
+    def right(self):
+        return self.__right
+
+    @right.setter
+    def right(self, value):
+        # since right is not just any variable but a node itself
+        if (isinstance(value, Node)) or (value is None):
+            self.__right = value
+        else:
+            print("Invalid right node, should be an instance of class 'Node'.")
+
+    def insert(self, value):
+        # no duplicate data
```

Comment:
I'm confused over this one that if this should be kept or not.
ping @singhpratyush @aviaryan @aashutoshrathi